jena-arq/src/main/java/org/apache/jena/system/Txn.java

```diff
@@ -137,4 +139,46 @@ private static <T extends Transactional> void onThrowable(Throwable th, T txn) {
             txn.end();
         } catch (Throwable th2) { th.addSuppressed(th2); }
     }
+
+    /**
+     * Begins a transaction and returns a transaction control instance suitable
+     * for use with try-with-resources blocks.
+     * See {@link #begin(Transactional, TxnType)}.
+     */
+    public static AutoTxn begin(Transactional txn, ReadWrite readWrite) {
+        return begin(txn, TxnType.convert(readWrite));
+    }
+
+    /**
+     * Begins a transaction and returns a transaction control instance suitable
+     * for use with try-with-resources blocks.
+     * Allows for raising checked exceptions in an idiomatic way.
+     * Closing the TxnCtl instance will abort the transaction unless it
+     * has been manually committed.
+     * <p>
+     *
+     * Usage example:
+     * <pre>
+     * public void myMethod() throws IOException {
+     *   try (AutoTxn txn = Txn.begin(dataset, TxnType.WRITE)) {
+     *     // Do work
+     *     if (someError) {
+     *         throw new IOException();
+     *     }
+     *     // Must manually call commit on success.
+     *     txn.commit();
```

Comment:
That's a different style to other `Txn` which commits on normal termination of the runnable and aborts on an exception, though I don't immediately see how to do the exception capture here.

Reply:
When `AutoTxn.close()` is called by the try-with-resources block, then there seems to be no standard way for close to detect whether it is called after normal or exceptional block exit. Therefore the logic is that always exceptional block exit is assumed (-> rollback) and commit() has to be called manually.